test/core/tsi/ssl_transport_security_utils_test.cc

```diff
@@ -0,0 +1,417 @@
+    SSL_CTX_set_min_proto_version(client_ctx, TLS1_3_VERSION);
+    SSL_CTX_set_min_proto_version(server_ctx, TLS1_3_VERSION);
+    SSL_CTX_set_max_proto_version(client_ctx, TLS1_3_VERSION);
+    SSL_CTX_set_max_proto_version(server_ctx, TLS1_3_VERSION);
+
+    // Create client and server connection objects and configure their BIOs.
+    SSL* client(SSL_new(client_ctx));
+    SSL* server(SSL_new(server_ctx));
+
+    SSL_CTX_free(client_ctx);
+    SSL_CTX_free(server_ctx);
+
+    // Turns off issuance of session tickets by servers.
+    SSL_set_options(client, SSL_OP_NO_TICKET);
+    SSL_set_options(server, SSL_OP_NO_TICKET);
+
+    SSL_set_connect_state(client);
+    SSL_set_accept_state(server);
+    BIO* bio1;
+    BIO* bio2;
+    BIO_new_bio_pair(&bio1, /*writebuf1=*/0, &bio2, /*writebuf2=*/0);
+    SSL_set_bio(client, bio1, bio1);
+    SSL_set_bio(server, bio2, bio2);
+
+    // Drive both the client and server handshake operations to completion.
+    while (true) {
+      int client_ret = SSL_do_handshake(client);
+      int client_err = SSL_get_error(client, client_ret);
+      if (client_err != SSL_ERROR_NONE && client_err != SSL_ERROR_WANT_READ &&
+          client_err != SSL_ERROR_WANT_WRITE &&
+          client_err != SSL_ERROR_PENDING_TICKET) {
+        return absl::InternalError(
+            absl::StrCat("Client error:", SSL_error_description(client_err)));
+      }
+
+      int server_ret = SSL_do_handshake(server);
+      int server_err = SSL_get_error(server, server_ret);
+      if (server_err != SSL_ERROR_NONE && server_err != SSL_ERROR_WANT_READ &&
+          server_err != SSL_ERROR_WANT_WRITE &&
+          server_err != SSL_ERROR_PENDING_TICKET) {
+        return absl::InternalError(
+            absl::StrCat("Server error:", SSL_error_description(server_err)));
+      }
+      if (client_ret == 1 && server_ret == 1) {
+        break;
+      }
+    }
+
+    *out_client = client;
+    *out_server = server;
+
+    return absl::OkStatus();
+  }
+
+  static std::size_t CalculateRecordSizeFromHeader(uint8_t fourth_header_byte,
+                                                   uint8_t fifth_header_byte) {
+    return (static_cast<int>(fourth_header_byte & 0xff) << 8) +
+           static_cast<int>(fifth_header_byte & 0xff);
+  }
+
+  void SetUp() override {
+    ASSERT_EQ(DoHandshake(&client_ssl, &server_ssl), absl::OkStatus());
+
+    ASSERT_THAT(client_ssl, NotNull());
+    ASSERT_THAT(server_ssl, NotNull());
+
+    BIO* client_ssl_bio = nullptr;
+    ASSERT_EQ(BIO_new_bio_pair(&client_bio, /*writebuf1=*/0, &client_ssl_bio,
+                               /*writebuf2=*/0),
+              1);
+    SSL_set_bio(client_ssl, client_ssl_bio, client_ssl_bio);
+
+    BIO* server_ssl_bio = nullptr;
+    ASSERT_EQ(BIO_new_bio_pair(&server_bio, /*writebuf1=*/0, &server_ssl_bio,
+                               /*writebuf2=*/0),
+              1);
+    SSL_set_bio(server_ssl, server_ssl_bio, server_ssl_bio);
+
```

Comment:
one small nit: can we please remove all the unnecessary blank spaces in this PR? If you intend the blank space to be a "logical breakpoint", you can add a comment explaining what the next block will do. Otherwise I think it's safe to just remove these blank spaces, and same thing for all the other files.